std/experimental/allocator/building_blocks/aligned_block_list.d

```diff
@@ -0,0 +1,690 @@
+        {
+            size_t bytesUsed;
+        }
+    }
+
+    // Root of the internal doubly linked list
+    AlignedBlockNode* root;
+
+    // Number of active nodes
+    int numNodes;
+
+    // If the numNodes exceeds this limit, we will start deallocating nodes
+    enum maxNodes = 50;
+
+    // This lock is always taken when changing the list
+    // To improve performance, the lock is not taken when the allocation logic is called
+    static if (isShared)
+    SpinLock lock = SpinLock(SpinLock.Contention.brief);
+
+    // Moves a node to the front of the list, allowing for quick allocations
+    void moveToFront(AlignedBlockNode* tmp)
+    {
+        auto localRoot = cast(AlignedBlockNode*) root;
+        if (tmp == localRoot)
+            return;
+
+        if (tmp.prev) tmp.prev.next = tmp.next;
+        if (tmp.next) tmp.next.prev = tmp.prev;
+        if (localRoot) localRoot.prev = tmp;
+        tmp.next = localRoot;
+        tmp.prev = null;
+
+        root = cast(typeof(root)) tmp;
+    }
+
+    // Removes a node from the list, including its payload
+    // The payload is deallocated by calling 'parent.deallocate'
+    void removeNode(AlignedBlockNode* tmp)
+    {
+        auto next = tmp.next;
+        if (tmp.prev) tmp.prev.next = tmp.next;
+        if (tmp.next) tmp.next.prev = tmp.prev;
+        assert(parent.deallocate((cast(void*) tmp)[0 .. alignment]));
+
+        if (tmp == cast(AlignedBlockNode*) root)
+            root = cast(typeof(root)) next;
+
+        static if (isShared)
+        {
+            import core.atomic : atomicOp;
+            atomicOp!"-="(numNodes, 1);
+        }
+        else
+        {
+            numNodes--;
+        }
+    }
+
+    // If the nodes do not have available space, a new node is created
+    // by drawing memory from the parent allocator with aligned allocations.
+    // The new node is inserted at the front of the list
+    bool insertNewNode()
+    {
+        void[] buf = parent.alignedAllocate(alignment, alignment);
+        if (buf is null)
+            return false;
+
+        auto localRoot = cast(AlignedBlockNode*) root;
+        auto newNode = cast(AlignedBlockNode*) buf;
+
+        // The first part of the allocation represent the node contents
+        // followed by the actual payload
+        ubyte[] payload = cast(ubyte[]) buf[AlignedBlockNode.sizeof .. $];
+        newNode.bAlloc = Allocator(payload);
+
+        newNode.next = localRoot;
+        newNode.prev = null;
+        if (localRoot)
+            localRoot.prev = newNode;
+        root = cast(typeof(root)) newNode;
+
+        static if (isShared)
+        {
+            import core.atomic : atomicOp;
+            atomicOp!"+="(numNodes, 1);
+        }
+        else
+        {
+            numNodes++;
+        }
+
+        return true;
+    }
+
+public:
+    static if (stateSize!ParentAllocator) ParentAllocator parent;
+    else alias parent = ParentAllocator.instance;
+
+    enum ulong alignment = theAlignment;
```

Comment:
I don't see how this could be correct if `theAlignment` is the size of each node and memory is allocated at an offset from the start of the node.

Reply:
Since the metadata for each node and the payload are adjacent in memory, I considered the alignment to be for the whole block (metadata + payload).
Inside the payload, an `Allocator` can manage the allocations with its own alignment.

Reply:
At the same time, I think it's not great for the caller to expect each allocation to have an alignment, but we silently add some padding before.
I think it's better for the alignment to be (metadata + payload), but clearly specify that and expose the length of the metadata.

Reply:
Because other allocators in `std.experiment.allocator` use `alignment` to specifically mean the alignment of memory returned by `allocate`, use a different name here and everything will be fine. Otherwise, templated code that uses design-by-introspection can leap to a wrong conclusion -- the compiler doesn't read the DDOC, so a comment will not prevent that!